lib/TPP/CombineTpp.cpp

```diff
@@ -0,0 +1,128 @@
+//===- CombineTpp.cpp --------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Tpp/TppOps.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+struct CombineTppOpPattern : public OpRewritePattern<tpp::ReluOp> {
+  using OpRewritePattern<tpp::ReluOp>::OpRewritePattern;
+
+  // TODO this is not a sufficient check.
+  static bool isVNNILayout(Operation *brgemmOp) {
+    if (!brgemmOp->getOpOperands()[0]
+             .get()
+             .getType()
+             .cast<ShapedType>()
+             .getElementType()
+             .isBF16())
+      return false;
+    return (brgemmOp->getOpOperands()[0]
+                    .get()
+                    .getType()
+                    .cast<ShapedType>()
+                    .getRank() +
+                1 ==
+            brgemmOp->getOpOperands()[1]
+                .get()
+                .getType()
+                .cast<ShapedType>()
+                .getRank());
+  }
+
+  LogicalResult matchAndRewrite(tpp::ReluOp reluOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto definingOp = reluOp->getOperands()[0].getDefiningOp();
+    Operation *addOp = NULL;
+    for (Operation *user : definingOp->getUsers()) {
+      if (user == definingOp || user == reluOp ||
+          isa<memref::DeallocOp>(user)) {
+        continue;
+      }
+      // assert(user.isa<tpp::AddBCastOp>());
+      addOp = user;
+      break;
+    }
+
+    if (addOp == NULL) {
+      return failure();
+    }
+    auto brgemmResultBuffer = addOp->getOperands()[0].getDefiningOp();
+    Operation *brgemmOp = NULL;
+    for (Operation *user : brgemmResultBuffer->getUsers()) {
+      if (isa<memref::CastOp>(user)) {
+        for (Operation *castUser : user->getUsers()) {
+          if (isa<tpp::BrgemmOp>(castUser) ||
+              isa<tpp::VNNIBrgemmOp>(castUser)) {
+            brgemmOp = castUser;
```

Comment:
Same here. We don't want to assume the shape of the graph to be "nice looking" always.

Reply:
We either assume that or we risk not matching patterns that match fused brgemm. In order to make the pattern matcher more robust, we should operate at tensor level, to avoid such checks entirely.